File: huggingplace-python-sdk/huggingplace_sdk/trace/sender.py
"""
Send trace data to HuggingPlace backend
"""

import asyncio
import time
from typing import Dict, Any, List, Optional, Callable
import requests
from .config import DEFAULT_TRACE_CONFIG


def create_sender(config: Dict[str, Any] = None) -> Dict[str, Callable]:
    """
    Create a configured sender instance
    
    Args:
        config: Configuration object
        
    Returns:
        Sender instance with methods
    """
    if config is None:
        config = {}
    
    sender_config = {**DEFAULT_TRACE_CONFIG, **config}
    
    # Build endpoints from base_url
    base_url = sender_config.get("base_url", DEFAULT_TRACE_CONFIG["base_url"])
    trace_endpoint = f"{base_url}/api/traces"
    batch_endpoint = f"{base_url}/api/traces/batch"
    
    # Instance state
    trace_batch = []
    batch_timeout = None

    async def send_trace_with_retry(trace_data: Dict[str, Any], retries: int = 0) -> Optional[Dict[str, Any]]:
        """
        Send a single trace with retry logic
        
        Args:
            trace_data: Trace data to send
            retries: Current retry attempt
            
        Returns:
            Response data or None if failed
        """
        try:
            response = requests.post(
                trace_endpoint,
                json=trace_data,
                timeout=sender_config["timeout"] / 1000,  # Convert to seconds
                headers=sender_config["headers"]
            )
            
            if response.status_code == 201:
                if not sender_config["silent"]:
                    print(f"✅ Trace sent successfully: {response.json().get('traceId')}")
                return response.json()
            else:
                raise Exception(f"Unexpected status: {response.status_code}")
        except Exception as error:
            if not sender_config["silent"]:
                print(f"❌ Failed to send trace (attempt {retries + 1}): {str(error)}")
            
            if retries < sender_config["max_retries"]:
                # Exponential backoff
                delay = sender_config["retry_delay"] * (2 ** retries)
                if not sender_config["silent"]:
                    print(f"🔄 Retrying in {delay}ms...")
                
                await asyncio.sleep(delay / 1000)  # Convert to seconds
                return await send_trace_with_retry(trace_data, retries + 1)
            else:
                if not sender_config["silent"]:
                    print("❌ Max retries reached, dropping trace")
                return None

    async def send_batch_with_retry(traces: List[Dict[str, Any]], retries: int = 0) -> Optional[Dict[str, Any]]:
        """
        Send a batch of traces with retry logic
        
        Args:
            traces: Array of trace data
            retries: Current retry attempt
            
        Returns:
            Response data or None if failed
        """
        try:
            response = requests.post(
                batch_endpoint,
                json=traces,
                timeout=sender_config["batch_timeout_ms"] / 1000,  # Convert to seconds
                headers=sender_config["headers"]
            )
            
            if response.status_code == 201:
                if not sender_config["silent"]:
                    print(f"✅ Batch sent successfully: {response.json().get('count')} traces")
                return response.json()
            else:
                raise Exception(f"Unexpected status: {response.status_code}")
        except Exception as error:
            if not sender_config["silent"]:
                print(f"❌ Failed to send batch (attempt {retries + 1}): {str(error)}")
            
            if retries < sender_config["max_retries"]:
                delay = sender_config["retry_delay"] * (2 ** retries)
                if not sender_config["silent"]:
                    print(f"🔄 Retrying batch in {delay}ms...")
                
                await asyncio.sleep(delay / 1000)  # Convert to seconds
                return await send_batch_with_retry(traces, retries + 1)
            else:
                if not sender_config["silent"]:
                    print("❌ Max retries reached, dropping batch")
                return None

    def flush_batch():
        """Flush the current batch of traces"""
        nonlocal trace_batch
        if trace_batch:
            batch = trace_batch.copy()
            trace_batch.clear()
            
            asyncio.create_task(send_batch_with_retry(batch)).add_done_callback(
                lambda task: print(f"❌ Failed to flush batch: {task.exception()}") if task.exception() else None
            )

    async def send_trace(trace: Dict[str, Any]) -> None:
        """
        Send a trace (single or batched)
        
        Args:
            trace: Trace data to send
        """
        try:
            # Validate required fields
            if not all(key in trace for key in ["traceId", "spanId", "operation"]):
                if not sender_config["silent"]:
                    print("❌ Invalid trace data: missing required fields")
                return

            # Add to batch if batching is enabled
            if sender_config["batch_size"] > 1:
                trace_batch.append(trace)
                
                # Clear existing timeout
                nonlocal batch_timeout
                if batch_timeout:
                    batch_timeout.cancel()
                
                # Flush if batch is full
                if len(trace_batch) >= sender_config["batch_size"]:
                    flush_batch()
                else:
                    # Set timeout to flush remaining traces
                    batch_timeout = asyncio.create_task(
                        asyncio.sleep(sender_config["batch_timeout"] / 1000)
                    )
                    batch_timeout.add_done_callback(lambda _: flush_batch())
            else:
                # Send immediately if batching is disabled
                await send_trace_with_retry(trace)
        except Exception as error:
            if not sender_config["silent"]:
                print(f"❌ Error in send_trace: {str(error)}")

    return {
        "send_trace": send_trace,
        "send_trace_with_retry": send_trace_with_retry,
        "send_batch_with_retry": send_batch_with_retry,
        "flush_batch": flush_batch
    }


# Create default sender instance for backward compatibility
_default_sender = create_sender()


async def send_trace(trace: Dict[str, Any]) -> None:
    """
    Send a trace using the default sender (backward compatibility)
    
    Args:
        trace: Trace data to send
    """
    return await _default_sender["send_trace"](trace)

File: huggingplace-sdk/src/trace/sender.js
// sender.js - Send trace data to HuggingPlace backend
import axios from 'axios';
import { DEFAULT_TRACE_CONFIG } from '../config.js';

/**
 * Create a configured sender instance
 * @param {Object} config - Configuration object
 * @returns {Object} Sender instance with methods
 */
export function createSender(config = {}) {
    const senderConfig = { ...DEFAULT_TRACE_CONFIG, ...config };
    
    // Build endpoints from baseUrl
    const baseUrl = senderConfig.baseUrl || DEFAULT_TRACE_CONFIG.baseUrl;
    const traceEndpoint = `${baseUrl}/api/traces`;
    const batchEndpoint = `${baseUrl}/api/traces/batch`;
    
    // Instance state
    let traceBatch = [];
    let batchTimeout = null;

    /**
     * Send a single trace with retry logic
     * @param {Object} traceData - Trace data to send
     * @param {number} retries - Current retry attempt
     * @returns {Promise<Object|null>} Response data or null if failed
     */
    async function sendTraceWithRetry(traceData, retries = 0) {
        try {
            const response = await axios.post(traceEndpoint, traceData, {
                timeout: senderConfig.timeout,
                headers: senderConfig.headers
            });
            
            if (response.status === 201) {
                if (!senderConfig.silent) {
                    console.log(`✅ Trace sent successfully: ${response.data.traceId}`);
                }
                return response.data;
            } else {
                throw new Error(`Unexpected status: ${response.status}`);
            }
        } catch (error) {
            if (!senderConfig.silent) {
                console.error(`❌ Failed to send trace (attempt ${retries + 1}):`, error.message);
            }
            
            if (retries < senderConfig.maxRetries) {
                // Exponential backoff
                const delay = senderConfig.retryDelay * Math.pow(2, retries);
                if (!senderConfig.silent) {
                    console.log(`🔄 Retrying in ${delay}ms...`);
                }
                
                await new Promise(resolve => setTimeout(resolve, delay));
                return sendTraceWithRetry(traceData, retries + 1);
            } else {
                if (!senderConfig.silent) {
                    console.error('❌ Max retries reached, dropping trace');
                }
                return null;
            }
        }
    }

    /**
     * Send a batch of traces with retry logic
     * @param {Array} traces - Array of trace data
     * @param {number} retries - Current retry attempt
     * @returns {Promise<Object|null>} Response data or null if failed
     */
    async function sendBatchWithRetry(traces, retries = 0) {
        try {
            const response = await axios.post(batchEndpoint, traces, {
                timeout: senderConfig.batchTimeoutMs,
                headers: senderConfig.headers
            });
            
            if (response.status === 201) {
                if (!senderConfig.silent) {
                    console.log(`✅ Batch sent successfully: ${response.data.count} traces`);
                }
                return response.data;
            } else {
                throw new Error(`Unexpected status: ${response.status}`);
            }
        } catch (error) {
            if (!senderConfig.silent) {
                console.error(`❌ Failed to send batch (attempt ${retries + 1}):`, error.message);
            }
            
            if (retries < senderConfig.maxRetries) {
                const delay = senderConfig.retryDelay * Math.pow(2, retries);
                if (!senderConfig.silent) {
                    console.log(`🔄 Retrying batch in ${delay}ms...`);
                }
                
                await new Promise(resolve => setTimeout(resolve, delay));
                return sendBatchWithRetry(traces, retries + 1);
            } else {
                if (!senderConfig.silent) {
                    console.error('❌ Max retries reached, dropping batch');
                }
                return null;
            }
        }
    }

    /**
     * Flush the current batch of traces
     */
    function flushBatch() {
        if (traceBatch.length > 0) {
            const batch = [...traceBatch];
            traceBatch = [];
            
            sendBatchWithRetry(batch).catch(error => {
                if (!senderConfig.silent) {
                    console.error('❌ Failed to flush batch:', error);
                }
            });
        }
    }

    /**
     * Send a trace (single or batched)
     * @param {Object} trace - Trace data to send
     */
    async function sendTrace(trace) {
        try {
            // Validate required fields
            if (!trace.traceId || !trace.spanId || !trace.operation) {
                if (!senderConfig.silent) {
                    console.error('❌ Invalid trace data: missing required fields');
                }
                return;
            }

            // Add to batch if batching is enabled
            if (senderConfig.batchSize > 1) {
                traceBatch.push(trace);
                
                // Clear existing timeout
                if (batchTimeout) {
                    clearTimeout(batchTimeout);
                }
                
                // Flush if batch is full
                if (traceBatch.length >= senderConfig.batchSize) {
                    flushBatch();
                } else {
                    // Set timeout to flush remaining traces
                    batchTimeout = setTimeout(flushBatch, senderConfig.batchTimeout);
                }
            } else {
                // Send immediately if batching is disabled
                await sendTraceWithRetry(trace);
            }
        } catch (error) {
            if (!senderConfig.silent) {
                console.error('❌ Error in sendTrace:', error);
            }
        }
    }

    return {
        sendTrace,
        sendTraceWithRetry,
        sendBatchWithRetry,
        flushBatch
    };
}

// Create default sender instance for backward compatibility
const defaultSender = createSender();

/**
 * Send a trace using the default sender (backward compatibility)
 * @param {Object} trace - Trace data to send
 */
export async function sendTrace(trace) {
    return await defaultSender.sendTrace(trace);
}

// Graceful shutdown - flush any remaining traces
process.on('SIGINT', () => {
    console.log('Flushing remaining traces...');
    defaultSender.flushBatch();
    process.exit(0);
});

process.on('SIGTERM', () => {
    console.log('Flushing remaining traces...');
    defaultSender.flushBatch();
    process.exit(0);
});
